---
Title: "Data Cleaning and Imputation by Austin Hayes"

Author:
  - Name: Austin Hayes

  -  Email: ahayes65@charlotte.edu

  -  Affiliation: University of North Carolina at Charlotte

Date: June 26, 2025

Description: Using NFL Scouting Combine Event scores from 2004 - 2023, we will learn about data cleaning and imputation in Python.

Categories:
  - Interpreting findings
  - Ethics
  - Importing and Reading data
  - Data Cleaning
  - Data Science
  - Pandas
  - Data Imputation


### Data

This Dataset is from the SCORE Network Data Repository. The authors include: Shane Hauk, Michael Schuckers and Robin Lock

Visit the original data page here: https://data.scorenetwork.org/football/nfl-draft-combine.html

The data set contains 6128 rows and 8 columns. Each row represents a player at the NFL Scouting Combine between 2004 and 2023.

Download data: 

Available on the [Intro to Data Cleaning and Imputation by Austin Hayes](https://github.com/schuckers/Charlotte_SCORE_Summer25/tree/main/Data%20for%20Modules/Data%20for%20Intro%20to%20Data%20Cleaning%20and%20Imputation%20by%20Austin%20Hayes): [epl_player_stats_24_25.csv](https://raw.githubusercontent.com/schuckers/Charlotte_SCORE_Summer25/refs/heads/main/Data%20for%20Modules/Data%20for%20Intro%20to%20Data%20Cleaning%20and%20Imputation%20by%20Austin%20Hayes/nfl_combine.csv)

---

### Variables and their Descriptions:


<details>
<summary><b>Variable Descriptions</b></summary>

| Variable | Description | 
|----|-------------|
| position | Playing position of the player |
| Round | Round player was drafted in |
| forty | 40-yard dash time |
| vertical | Vertical jump height (inches) |
| bench_reps | 225 bench press reps |
| broad_jump | Broad jump distances (inches) |
| shuttle | 20-yard shuttle time |

</details>

---

## Learning Goals

- Learn about the 3 C's of Data cleaning
- Basic principles of Data Cleaning
- Data Imputation
- Use Pandas for Data Cleaning
- Learning about the Ethics of Data Cleaning

---

# Data Cleaning and the "Three C's"

Data cleaning is an essential part of data science and sports analytics. Most of the time, you need to clean up the dataset before using it because it won't always be in a correct or readable format. This brings us to the "three C's". This is a very important aspect of data science that provides basic guidance on what to look for when undergoing the data cleaning process. 

The three C's stand for:
- Consistent
- Complete
- Correct


_Consistent_: This means that your dataset is correctly formatted in a standardized fashion. All instances are properly formatted into their respective data types and there are no inconsistencies with text casing.

_Complete_: There is no missing data. All missing values have been addressed by filling them with a reasonable substitute, imputation (we will discuss this later) and/or removing the data depending on the context (we will discuss the ethics of this later).

_Correct_: Fixing inconsistent data entrys such as misspelled items, inconceivable outliers and/or ilogical errors such as negative ages.


Let's break these down by diving into the football combine data!

---

# Getting Started

Before we get started let's import pandas so we can read in our data.

NOTE: If one of the import statements does not work, you may need to download the library(s). Visit one of the links below for more information on downloading them.

Pandas: https://pandas.pydata.org/docs/getting_started/install.html 

Numpy: https://numpy.org/install/

Now, lets import our libraries:

In [61]:
# Import necessary libraries

#Import numpy library for numerical operations
import numpy as np

#Import pandas library for data manipulation and analysis
import pandas as pd

In [62]:
# Read the in the NFL Combine data and store it in a DataFrame
combine_data = pd.read_csv('https://raw.githubusercontent.com/schuckers/Charlotte_SCORE_Summer25/refs/heads/main/Data%20for%20Modules/Data%20for%20Intro%20to%20Data%20Cleaning%20and%20Imputation%20by%20Austin%20Hayes/nfl_combine.csv')

---

# Consistency and Correctness

Let's start by examining the first 5 rows of data to get a better look.

In [63]:
# Reveal first 5 rows of the dataset
combine_data.head()

position  Round  forty  vertical  bench_reps  broad_jump  three_cone  \
0       QB    NaN   4.79      30.5         NaN       110.0        7.66   
1       RB    NaN   4.50      34.0        21.0       121.0        7.09   
2       QB    NaN   4.60       NaN         NaN         NaN         NaN   
3       QB    NaN   4.95      30.0         NaN       119.0        7.44   
4       WR    NaN   4.78      38.0         NaN       118.0         NaN   

   shuttle  
0     4.41  
1     4.30  
2      NaN  
3     4.34  
4     4.45

Interesting! Looks like in just the first few rows, there are a lot of missing values.

Before we start deciding what to do with these missing values, let's start by checking the consistentcy. Let's check the datatypes for column and make sure that they align with their values.

In [64]:
# Check the data types of each column to ensure they are consistent with the expected data types
combine_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6128 entries, 0 to 6127
Data columns (total 8 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   position    6128 non-null   object 
 1   Round       4023 non-null   float64
 2   forty       5751 non-null   float64
 3   vertical    4965 non-null   float64
 4   bench_reps  4308 non-null   float64
 5   broad_jump  4891 non-null   float64
 6   three_cone  3985 non-null   float64
 7   shuttle     4093 non-null   float64
dtypes: float64(7), object(1)
memory usage: 383.1+ KB


Nice! looks like we don't need to change any of the datatypes for this dataset.

Let's go ahead and check the _correctness_ and consistency of the dataset by checking some of the rows. To do this, let's take a look at the first 50 rows.

After running the code block below, examine the dataset to make sure they are correctly formatted and there are no absurd outliers. We also need to check that there are no misspelled positions or capitalization issues. 

In [65]:
# Let's check the first 50 rows of the dataset to ensure correctness and consistency.
# This will help us identify any potential issues with the data formatting or inconsistencies.
combine_data.head(50)

position  Round  forty  vertical  bench_reps  broad_jump  three_cone  \
0        QB    NaN   4.79      30.5         NaN       110.0        7.66   
1        RB    NaN   4.50      34.0        21.0       121.0        7.09   
2        QB    NaN   4.60       NaN         NaN         NaN         NaN   
3        QB    NaN   4.95      30.0         NaN       119.0        7.44   
4        WR    NaN   4.78      38.0         NaN       118.0         NaN   
5        OT    NaN   5.41      28.0        20.0       104.0        8.24   
6        WR    NaN   4.76      31.5         NaN       116.0        7.41   
7        WR    NaN   4.73      38.0         NaN       128.0        7.10   
8        WR    NaN   4.59      28.5         NaN       111.0        7.24   
9        RB    NaN   4.71      31.5        16.0       108.0         NaN   
10       QB    NaN   4.87      26.5         NaN       110.0        7.69   
11       WR    NaN   4.61      32.5         NaN       118.0        7.16   
12       WR    NaN   4.50       NaN         NaN         NaN         NaN   
13        C    NaN   4.98      30.5        28.0       105.0        7.84   
14       OG    NaN   5.67      23.0        18.0        94.0        8.25   
15       OG    NaN   5.54      26.5        27.0       101.0        8.12   
16       QB    NaN   4.83      30.5         NaN       113.0        7.28   
17       RB    NaN   4.68      34.5        25.0       120.0        7.44   
18       OG    NaN   5.50      26.5        21.0        98.0        8.01   
19       RB    NaN   4.54      33.5        13.0       112.0        7.27   
20       WR    NaN   4.60      31.0         NaN       117.0        6.72   
21       RB    NaN   4.83      30.5        14.0       108.0         NaN   
22       OG    NaN   6.05       NaN         NaN         NaN        8.03   
23       WR    NaN   4.69      35.0         NaN       112.0         NaN   
24       QB    NaN   4.95       NaN         NaN         NaN         NaN   
25       OG    NaN   5.10       NaN        32.0         NaN         NaN   
26        C    NaN   5.25       NaN        19.0        85.0        8.31   
27       WR    NaN   4.56      33.0         NaN       117.0        7.38   
28       TE    NaN   4.89       NaN        12.0       102.0         NaN   
29       WR    NaN   4.46      31.0         NaN       113.0         NaN   
30        C    NaN   5.84      24.0         NaN        93.0        8.09   
31       OG    NaN   5.82      25.5        21.0        93.0        8.38   
32       FB    NaN   4.97      32.5        22.0       114.0        7.50   
33       TE    NaN   4.84      30.0        19.0       106.0        7.51   
34       WR    NaN   4.62      37.0         NaN       121.0        7.06   
35       TE    NaN   5.03       NaN        21.0       114.0        7.33   
36       RB    NaN   4.70      30.5        21.0       115.0        7.25   
37       WR    NaN   4.65      34.0         NaN       122.0        7.12   
38       RB    NaN   4.64      32.0         NaN       114.0        7.41   
39       TE    7.0   4.89      30.0        17.0       106.0        7.86   
40       OT    1.0   5.49      25.0        27.0        98.0        8.25   
41       OT    4.0   5.07      30.5        34.0       106.0        8.09   
42       OG    5.0   5.22       NaN        26.0         NaN         NaN   
43       RB    2.0   4.37      38.5        25.0       117.0         NaN   
44       WR    3.0   4.58      38.0         NaN       128.0        7.34   
45       OT    6.0   5.11      30.5         NaN       105.0        7.96   
46       OG    1.0   5.33       NaN        30.0         NaN         NaN   
47       WR    5.0   4.36       NaN         NaN         NaN         NaN   
48       RB    NaN   4.51      34.5        22.0       112.0         NaN   
49       OT    6.0   5.44      27.0        22.0       101.0        7.56   

    shuttle  
0      4.41  
1      4.30  
2       NaN  
3      4.34  
4      4.45  
5      4.61  
6       NaN  
7      4.09  
8      4.27  
9       NaN  
10     4.22  
11     4.

Looks quite good! There are no capitalization issues in the position column and all of the numeric values are in float (decimal) format.

This is a massive dataset with over 6000 rows so it may be best to examine it using .describe() from the Pandas library. This will allow us to see statistical information about each column, which will tell us if there are any alarming outliers.

In [66]:
# Check the statistical summary of the dataset to identify any outliers or anomalies
combine_data.describe()

Round        forty     vertical   bench_reps   broad_jump  \
count  4023.000000  5751.000000  4965.000000  4308.000000  4891.000000   
mean      3.832712     4.764519    32.838933    20.721681   114.984870   
std       1.931859     0.305037     4.259617     6.394058     9.325569   
min       1.000000     4.220000    17.500000     2.000000    74.000000   
25%       2.000000     4.530000    30.000000    16.000000   109.000000   
50%       4.000000     4.680000    33.000000    21.000000   116.000000   
75%       5.000000     4.960000    36.000000    25.000000   121.000000   
max       7.000000     6.050000    46.500000    49.000000   147.000000   

        three_cone      shuttle  
count  3985.000000  4093.000000  
mean      7.265425     4.402817  
std       0.406290     0.263690  
min       6.280000     3.750000  
25%       6.970000     4.210000  
50%       7.170000     4.360000  
75%       7.510000     4.560000  
max       9.040000     5.560000

This looks very good; none of the max/min times seem unrealistic. The only number that stands out to me is the bench rep max of 49 reps. This means that somebody holds the record for 49 reps of 225 bench; that's insane! 

FUN FACT: Stephen Paea is the player who had those 49 bench reps at the combine. He was a DT out of Oregan State and was drafted 53rd overall to the Chicago Bears in 2011.

Finally, we need to examine the completeness of the dataset. Before we address this, we need to talk about the ethical decision-making involved with missing data.

---


# Ethics of Data Completeness



We need to consider WHY the data is missing in the first place. Did they mean for it to be empty? Did they forget to record certain datapoints? Sometimes it can be extremely hard to tell so it is crucial to document some of the assumptions your making. Transparency is key for your interpertations so that users/readers understand your thinking process, which may affect the usage of the dataset. Context is everything with data completeness. With that being said, let's start tackling this dataset.



# Completeness

Building upon the previous sections, there are severl ways to approach this issue but let's start by asking: Why is the data missing? If it was intentionally done, why? 

In this case, any mising data is likely due to non-particpation. In more recent years, more and more athletes have started to skip certain drills or the combine altogether. It also depends on what variable we are refering to. For example, most of the missing numerical variables in this case are more than likely non=participation related. However, missing 'Round' numbers most likely means that specific player went undrafted. Missing 'position' values are a little harder to interpret. It may be because a player had no official position coming out of the combine and/or they may have played multiple positions in college. A perfect example of this was Travis Hunter in this year's draft because he had played both Cornerback and Wide Receiver in his college career. In this case, we may want to look at how many rows have missing positions. 

Before getting to that, let's make it simple and split the data. We will make one dataframe of missing values and the other will have complete records only.

In [67]:
# Make a copy of the dataset with no missing values
complete_combine_data = combine_data.dropna()

# Make a dataframe of the missing values
missing_combine_data = combine_data[combine_data.isnull().any(axis=1)]

Let's take a look at the volume of data for each dataframe.

In [68]:
# Display the shapes of the complete and missing dataframes to understand the extent of missing data
complete_combine_data.shape, missing_combine_data.shape

((1930, 8), (4198, 8))

That is a TON of data with missing values. Let's start by examining the rows with any missing position groups.

In [69]:
missing_combine_data[missing_combine_data['position'].isnull()]

Empty DataFrame
Columns: [position, Round, forty, vertical, bench_reps, broad_jump, three_cone, shuttle]
Index: []

That's great! There are no players with missing positions! However, let's just double check that all of them are properly formatted.

In [70]:
# Search every instance of a unique value in order to ensure that all positions are properly formatted
missing_combine_data['position'].unique()

array(['QB', 'RB', 'WR', 'OT', 'C', 'OG', 'TE', 'FB', 'OL', 'OLB', 'CB',
       'DT', 'S', 'DE', 'ILB', 'DL', 'EDGE', 'LB', 'DB'], dtype=object)

Looks good! Moving on to the numeric variables, let's see if there are any rows missing AT LEAST 4 of their combine drill values.

In [74]:
missing_combine_data[missing_combine_data[['forty','vertical','bench_reps','broad_jump','three_cone','shuttle']].isna().sum(axis=1) >= 4]

position  Round  forty  vertical  bench_reps  broad_jump  three_cone  \
2          QB    NaN   4.60       NaN         NaN         NaN         NaN   
12         WR    NaN   4.50       NaN         NaN         NaN         NaN   
24         QB    NaN   4.95       NaN         NaN         NaN         NaN   
25         OG    NaN   5.10       NaN        32.0         NaN         NaN   
42         OG    5.0   5.22       NaN        26.0         NaN         NaN   
...       ...    ...    ...       ...         ...         ...         ...   
6118       CB    7.0   4.51       NaN        18.0         NaN         NaN   
6122       DT    4.0   4.89       NaN         NaN         NaN         NaN   
6124       CB    5.0   4.26      42.0         NaN         NaN         NaN   
6125       DE    3.0    NaN       NaN        15.0         NaN         NaN   
6126       CB    NaN   4.57       NaN         NaN         NaN         NaN   

      shuttle  
2         NaN  
12        NaN  
24        NaN  
25        NaN  
42        NaN  
...       ...  
6118      NaN  
6122      NaN  
6124      NaN  
6125      NaN  
6126      NaN  

[1123 rows x 8 columns]

Wow, there are 1,123 players who had no data for the majority of the combine drills! In this case, these are most likely players that skipped out on certain drills in attempt to prevent injury and/or only do drills that may be neccesary for their position. It is difficult to decide what to do with these players.

Some of the possible options we could take are: drop all rows with missing data, drop ONLY the rows with majority of their data missing, mark the missing data as a certain value, OR change all of the missing drill values to the mean/median. 

Making this decision means we must weigh the pros and cons of each approach for the most faithful representation of the data. You would most likely not want to drop all of rows with missing values in this case because that is the majority of the dataset and causes an obstruction of the truth due to the elimination of a lot of drill values. Dropping players that have not competed in most of the drills could be a good idea but that would mean ignoring the drills that they DID do. On one hand, if you replace all of the data with a certain value such as the mean, it would'nt change the distribution too much. However, on the other hand, it would not be a truly accurate representation because if those players DID particpate in the drill, the outcome may be drastically different. 

Believe it or not, we may need to do several of these approaches. The best approach may be a somewhat subjective one but let's go with this one:

- To address assumpted undrafted players, we will change the column data type to the object type and enter all missing values as undrafted. 
- Players Missing ALL of their drill values will be dropped from the dataset since we only care about players that participated in the draft combine
- Any other missing drill value will be assigned to the mean value of that column in order to perserve drill values that were completed.


Ideally, we would not want to replace all of the missing values with the mean because it does not accurately represent those player TRUE drill results. However, in order to perserve an accurate datatype, float, this is most likely to be the best decision. The reason we changed the datatype of 'round' and not the drill value columns is due to the nature of 'round'. The round a player was drafted in is NOT a measurement. It can be viewed as a categorical variable, much like that of a shoe size. 'Round' may be represented by a number but it is read as a category that indicates when a player was drafted.

Let's start by dropping all players who, we assume, did not particpated in the draft combine.

In [60]:
# Drop rows that are missing all of the combine drill values
missing_combine_data = missing_combine_data.dropna(subset=['forty', 'vertical','bench_reps','broad_jump','three_cone','shuttle'], how='all')

# Check that they are no longer present
missing_combine_data[missing_combine_data[['forty', 'vertical','bench_reps','broad_jump','three_cone','shuttle']].isnull().all(axis=1)]

Empty DataFrame
Columns: [position, Round, forty, vertical, bench_reps, broad_jump, three_cone, shuttle]
Index: []

Nice! Now, let's change the data type of the round drafted column and replace all missing values with 'undrafted'

In [ ]:
# Change the column data type to the object type
missing_combine_data['Round'] = missing_combine_data['Round'].astype('object')

# Check that the data type has been changed
missing_combine_data.info()

C:\Users\austy\AppData\Local\Temp\ipykernel_16180\2563894076.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  missing_combine_data['Round'] = missing_combine_data['Round'].astype('object')


Looks like it is now an object datatype, now replace the missing values.

In [78]:
# Replace the missing values in the 'Round' column with 'undrafted'
missing_combine_data['Round'] = missing_combine_data['Round'].fillna('Undrafted')

C:\Users\austy\AppData\Local\Temp\ipykernel_16180\1807703531.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  missing_combine_data['Round'] = missing_combine_data['Round'].fillna('Undrafted')


In [79]:
# Let's check a sample of the data to ensure that the missing values have been replaced correctly
missing_combine_data.sample(50)

position      Round  forty  vertical  bench_reps  broad_jump  three_cone  \
777        FB        6.0   4.68      32.0         NaN       122.0        7.01   
3195       OT  Undrafted   4.88       NaN         NaN         NaN         NaN   
1262       OT  Undrafted   5.44      32.0        24.0       104.0        7.77   
372        QB  Undrafted   4.57      35.0         NaN       122.0        7.41   
2364       QB  Undrafted   4.81      29.5         NaN       112.0        7.18   
2165       OT        1.0   5.20      28.5         NaN         NaN        8.03   
5807       LB  Undrafted   4.38      39.0        24.0       133.0        7.14   
2702       TE        3.0   4.73      30.5        16.0       115.0         NaN   
4529        S  Undrafted   4.54      32.5        19.0       119.0        7.21   
3079       WR        4.0   4.32      39.0         NaN       135.0        6.65   
1762        C  Undrafted   5.15      27.0        27.0       113.0        7.97   
3748      OLB  Undrafted   4.77      29.5        18.0       113.0         NaN   
5402       CB  Undrafted   4.44      35.0        15.0       123.0         NaN   
4520       CB  Undrafted   4.63       NaN         7.0         NaN         NaN   
1985       WR        4.0   4.67      35.5         NaN       117.0        6.93   
2433       OT  Undrafted   4.90       NaN         NaN         NaN         NaN   
527        OG  Undrafted   5.46      25.5        28.0        98.0        7.80   
1026        C  Undrafted   5.40      28.5        22.0        99.0        8.15   
5153      ILB        3.0   4.72      31.5         NaN       115.0        6.73   
1376       QB  Undrafted   4.77      36.0        14.0       121.0        7.17   
2550       WR        2.0   4.31      40.0        11.0       135.0         NaN   
2632       WR  Undrafted   4.63      37.0        13.0       118.0        7.07   
1837       TE        2.0   4.75       NaN        20.0         NaN         NaN   
5016        S        2.0   4.52       NaN        16.0         NaN         NaN   
3885        S        4.0   4.39      31.5        16.0       110.0         NaN   
3021       WR  Undrafted   4.68      29.5        12.0       114.0        7.09   
300        WR        4.0   4.50       NaN         NaN         NaN         NaN   
3171       TE        4.0    NaN       NaN         NaN         NaN         NaN   
1364       TE  Undrafted   4.76       NaN         NaN         NaN         NaN   
5563       LB  Undrafted   4.62      32.5        23.0       123.0         NaN   
509        QB        1.0   4.48       NaN         NaN         NaN         NaN   
2823       RB  Undrafted   4.56      30.0        29.0       114.0        7.40   
2949       WR  Undrafted   4.54      40.0        15.0       132.0        6.91   
3338       DE        5.0   4.65      33.0         NaN       118.0         NaN   
4793        S  Undrafted   4.65      32.5        12.0       119.0        7.15   
5596       CB        4.0    NaN       NaN         NaN         NaN         NaN   
5847       LB  Undrafted   4.62      31.5        21.0       115.0        7.08   
4154      OLB  Undrafted   4.76      34.5        22.0       117.0        7.13   
136        OT        7.0   5.33       NaN        29.0         NaN         NaN   
2948       TE  Undrafted   4.78      31.5         NaN       119.0         NaN   
4222       DT        3.0   4.98      33.0         NaN        98.0        7.56   
5778        S        2.0   4.45      36.0         NaN       124.0         NaN   
84         WR  Undrafted   4.60      35.0         NaN       123.0        7.47   
4627      OLB        1.0   4.54      32.5         NaN       122.0        7.02   
4275       CB  Undrafted   4.54      33.5        15.0       120.0        7.23   
33         TE  Undrafted   4.84      30.0        19.0       106.0        7.51   
6115       LB        1.0   4.52      32.0         NaN       122.0         NaN   
26          C  Undrafted   5.25       NaN        19.0        85.0        8.31   
736        WR        3.0   4.35     

Looks great! Now we need to enter the rest of the missing values as the mean of their respective columns.

In [80]:
# Replace all values in the 'forty', 'vertical', 'bench_reps', 'broad_jump', 'three_cone', and 'shuttle' columns that are missing with the mean of their respective columns
# This will help us maintain the integrity of the data while filling in the gaps.

# 'forty' column
missing_combine_data['forty'] = missing_combine_data['forty'].fillna(missing_combine_data['forty'].mean())

# 'vertical' column
missing_combine_data['vertical'] = missing_combine_data['vertical'].fillna(missing_combine_data['vertical'].mean())

# 'bench_reps' column
missing_combine_data['bench_reps'] = missing_combine_data['bench_reps'].fillna(missing_combine_data['bench_reps'].mean())

# 'broad_jump' column
missing_combine_data['broad_jump'] = missing_combine_data['broad_jump'].fillna(missing_combine_data['broad_jump'].mean())

# 'three_cone' column
missing_combine_data['three_cone'] = missing_combine_data['three_cone'].fillna(missing_combine_data['three_cone'].mean())

# 'shuttle' column
missing_combine_data['shuttle'] = missing_combine_data['shuttle'].fillna(missing_combine_data['shuttle'].mean())

C:\Users\austy\AppData\Local\Temp\ipykernel_16180\2853139143.py:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  missing_combine_data['forty'] = missing_combine_data['forty'].fillna(missing_combine_data['forty'].mean())
C:\Users\austy\AppData\Local\Temp\ipykernel_16180\2853139143.py:8: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  missing_combine_data['vertical'] = missing_combine_data['vertical'].fillna(missing_combine_data['vertical'].mean())
C:\Users\austy\AppData\Local\Temp\ipykernel_16180\2853139143.py

Let's check a sample of the data to take a look at the results of our operations.

In [81]:
missing_combine_data.sample(50)

position      Round     forty   vertical  bench_reps  broad_jump  \
3887      OLB        2.0  4.710000  30.500000   20.044996  118.000000   
1606       TE        7.0  4.970000  32.733213    3.000000  115.086795   
891        OG  Undrafted  5.250000  32.733213   20.044996  115.086795   
3460       CB        3.0  4.520000  38.000000   20.044996  120.000000   
5412      OLB        2.0  4.700000  32.733213   17.000000  115.086795   
2811       TE        4.0  4.490000  32.733213   20.044996  115.086795   
679        WR  Undrafted  4.700000  32.500000   12.000000  114.000000   
1246       OT        2.0  5.410000  32.733213   30.000000  115.086795   
5936       DL        1.0  4.560000  36.000000   21.000000  125.000000   
5361      OLB  Undrafted  4.850000  29.500000   18.000000  112.000000   
211        FB  Undrafted  5.000000  25.500000   19.000000  100.000000   
824        TE        4.0  4.750000  32.733213   20.000000  115.086795   
712        FB  Undrafted  4.660000  39.000000   30.000000  134.000000   
5312       CB  Undrafted  4.490000  38.000000   13.000000  123.000000   
348        OG  Undrafted  5.570000  32.733213   21.000000  115.086795   
3193       OG  Undrafted  5.250000  22.500000   20.044996  100.000000   
4031       CB        5.0  4.460000  33.500000   20.044996  118.000000   
3106       OT        6.0  5.080000  28.500000   20.044996  109.000000   
2853       WR  Undrafted  4.450000  36.500000   20.044996  125.000000   
5993     EDGE        6.0  4.360000  37.000000   20.044996  131.000000   
1289       QB        1.0  4.510000  35.000000   20.044996  120.000000   
2762       WR        5.0  4.430000  38.500000   20.044996  122.000000   
2330       TE        5.0  4.762214  33.000000   22.000000  121.000000   
2203       WR  Undrafted  4.420000  34.500000    8.000000  123.000000   
2679       OT  Undrafted  5.270000  30.500000   23.000000  111.000000   
1575       RB        6.0  4.610000  34.000000   20.044996  122.000000   
372        QB  Undrafted  4.570000  35.000000   20.044996  122.000000   
1813       TE  Undrafted  4.790000  33.000000   24.000000  114.000000   
1919       WR        1.0  4.420000  33.000000   20.044996  120.000000   
3204       OT        1.0  5.020000  27.500000   20.044996  105.000000   
4488        S  Undrafted  4.670000  35.000000   20.000000  122.000000   
5897        S        2.0  4.480000  35.500000   19.000000  126.000000   
4657       DE        4.0  4.780000  32.733213   21.000000  115.086795   
1311       QB  Undrafted  4.590000  40.000000   20.044996  126.000000   
4592       DT        5.0  5.290000  32.733213   20.044996  115.086795   
2452       WR        5.0  4.500000  34.500000    8.000000  112.000000   
5678       DL        3.0  5.020000  27.500000   24.000000  115.086795   
3702        S  Undrafted  4.540000  32.733213   25.000000  115.086795   
5801       CB  Undrafted  4.570000  33.000000   12.000000  116.000000   
1844        C  Undrafted  5.170000  27.500000   22.000000  105.000000   
5909       DB        2.0  4.340000  35.000000   20.044996  128.000000   
6118       CB        7.0  4.510000  32.733213   18.000000  115.086795   
5518       CB  Undrafted  4.570000  37.000000   14.000000  125.000000   
1649       QB        7.0  4.760000  32.733213   20.044996  115.086795   
2659       QB  Undrafted  4.910000  33.000000   20.044996  116.000000   
2735       RB        2.0  4.530000  41.000000   23.000000  131.000000   
4537       DE  Undrafted  4.870000  35.500000   26.000000  116.000000   
1838       QB  Undrafted  4.660000  34.000000   20.044996  116.000000   
1248       OT  Undrafted  5.540000  26.500000   25.000000   99.000000   
6112       LB        3.0  4.470000  42.000000   20.044996  129.000000   

      three_cone   shuttle  
3887    7.010000  4.350000  
1606    7.275134  4.411082  
891     7.275134  4.411082  
3460    7.275134  3.900000  
5412    7.275134  4.411082  
2811    7.275134  4.411082  
679     7.040000  4.400000  
1246    7.275134  4.411082  
5936    7.275134  4.

Nice! Finally, let's go ahead and double-check that there are no more rows with missing values!

In [82]:
# Count up missing values in each column to ensure that all missing values have been addressed
missing_combine_data.isna().sum()

position      0
Round         0
forty         0
vertical      0
bench_reps    0
broad_jump    0
three_cone    0
shuttle       0
dtype: int64

Beatiful! Now let's complete the final step of this process and combine the complete and missing datasets back together!

In [83]:
# Lets combine the complete and missing datasets back together
Final_combine_data = pd.concat([complete_combine_data, missing_combine_data], ignore_index=True)

Let's check it out!

In [84]:
Final_combine_data.tail(50)

position      Round     forty   vertical  bench_reps  broad_jump  \
6078       LB        5.0  4.762214  32.733213   21.000000  115.086795   
6079        S  Undrafted  4.762214  35.000000   16.000000  120.000000   
6080        S  Undrafted  4.520000  32.733213   20.044996  124.000000   
6081       LB  Undrafted  4.760000  40.500000   20.044996  133.000000   
6082       DT        5.0  4.762214  32.733213   29.000000  115.086795   
6083     EDGE        2.0  4.550000  35.000000   20.044996  122.000000   
6084        S  Undrafted  4.762214  35.000000   15.000000  125.000000   
6085       DE        5.0  4.762214  32.733213   20.044996  115.086795   
6086       DE        2.0  4.770000  37.500000   20.044996  123.000000   
6087       DT  Undrafted  5.300000  27.000000   20.044996  100.000000   
6088        S        2.0  4.762214  35.000000   16.000000  115.086795   
6089       DT  Undrafted  4.762214  32.733213   20.044996  115.086795   
6090       DT        5.0  5.300000  32.733213   20.044996  101.000000   
6091     EDGE        5.0  4.720000  41.000000   25.000000  121.000000   
6092       LB        6.0  4.520000  39.500000   20.044996  120.000000   
6093       LB  Undrafted  4.762214  34.000000   20.044996  121.000000   
6094       LB  Undrafted  4.790000  31.000000   20.044996  109.000000   
6095       LB  Undrafted  4.730000  34.500000   20.000000  114.000000   
6096     EDGE        3.0  4.670000  33.000000   20.044996  120.000000   
6097       LB  Undrafted  4.670000  32.000000   20.044996  120.000000   
6098       LB        4.0  4.520000  37.500000   20.044996  128.000000   
6099     EDGE        5.0  4.860000  34.000000   20.044996  117.000000   
6100       LB        7.0  4.600000  36.000000   20.044996  122.000000   
6101       CB  Undrafted  4.480000  37.500000   20.044996  126.000000   
6102       CB        1.0  4.762214  32.733213   20.044996  115.086795   
6103       DE  Undrafted  4.810000  29.000000   20.044996  115.000000   
6104       CB        2.0  4.360000  34.500000   20.044996  128.000000   
6105       CB        2.0  4.380000  32.733213   20.044996  115.086795   
6106       CB  Undrafted  4.420000  32.733213   20.044996  123.000000   
6107     EDGE        1.0  4.580000  32.733213   27.000000  115.086795   
6108       DE        3.0  4.762214  32.733213   24.000000  115.086795   
6109     EDGE        7.0  4.700000  31.500000   20.044996  117.000000   
6110       CB  Undrafted  4.400000  32.733213   20.044996  115.086795   
6111        S  Undrafted  4.550000  34.000000   17.000000  122.000000   
6112       LB        3.0  4.470000  42.000000   20.044996  129.000000   
6113        S        5.0  4.470000  32.733213   20.044996  122.000000   
6114       DT        4.0  4.762214  33.000000   20.044996  107.000000   
6115       LB        1.0  4.520000  32.000000   20.044996  122.000000   
6116       DE        1.0  4.510000  35.500000   20.044996  123.000000   
6117       CB  Undrafted  4.680000  32.733213   20.044996  115.086795   
6118       CB        7.0  4.510000  32.733213   18.000000  115.086795   
6119       CB        4.0  4.762214  34.500000   17.000000  118.000000   
6120       CB        4.0  4.530000  36.000000   20.044996  124.000000   
6121     EDGE        2.0  4.460000  32.500000   20.044996  123.000000   
6122       DT        4.0  4.890000  32.733213   20.044996  115.086795   
6123        S        3.0  4.360000  39.500000   20.044996  128.000000   
6124       CB        5.0  4.260000  42.000000   20.044996  115.086795   
6125       DE        3.0  4.762214  32.733213   15.000000  115.086795   
6126       CB  Undrafted  4.570000  32.733213   20.044996  115.086795   
6127       DT        1.0  4.770000  29.000000   20.044996  111.000000   

      three_cone   shuttle  
6078    7.275134  4.411082  
6079    7.275134  4.411082  
6080    7.000000  4.411082  
6081    7.090000  4.550000  
6082    7.275134  4.411082  
6083    7.275134  4.450000  
6084    6.890000  4.411082  
6085    7.275134  4.411082  
6086    7.275134  4.

In [90]:
# Examine the column data types of the final combined dataset to ensure they are consistent
Final_combine_data.info()

# print the statistical summary of the final combined dataset
print(Final_combine_data.describe())

# Print the shape of the final combined dataset to confirm that it has been successfully created
print(Final_combine_data.shape)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6128 entries, 0 to 6127
Data columns (total 8 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   position    6128 non-null   object 
 1   Round       6128 non-null   object 
 2   forty       6128 non-null   float64
 3   vertical    6128 non-null   float64
 4   bench_reps  6128 non-null   float64
 5   broad_jump  6128 non-null   float64
 6   three_cone  6128 non-null   float64
 7   shuttle     6128 non-null   float64
dtypes: float64(6), object(2)
memory usage: 383.1+ KB
             forty     vertical   bench_reps   broad_jump   three_cone  \
count  6128.000000  6128.000000  6128.000000  6128.000000  6128.000000   
mean      4.764377    32.818868    20.520707   115.005445     7.268820   
std       0.295504     3.834316     5.369840     8.331267     0.327654   
min       4.220000    17.500000     2.000000    74.000000     6.280000   
25%       4.540000    31.000000    18.000000   111.000000   

In [91]:
# Confirm there are no missing values in the final combined dataset
print(Final_combine_data.isnull().sum())

position      0
Round         0
forty         0
vertical      0
bench_reps    0
broad_jump    0
three_cone    0
shuttle       0
dtype: int64


The data, much like the title of this section, is now 'complete'!

---


# Data Imputation


Data Imputation is an extremely important topic in data cleaning. YOu may not know what it is yet, but you have actually seen it in this module thus far!

_Data Imputation_: The process of filling in missing data points using estimated/substitute values. 

Do you remember when we replaced all of the missing combine drill values with each columns mean? That's data imputation!

However, that is not the only way to do data imputation. You can use different models to estimate what values to impute for your missing values. If you are new to data modeling or have no experience with it, I'll walk you through it. Model-making is just a way of using pre-existing data to estimate what a value might be. There are seperate models that can be used for categorical and numerical estimates. Today, we will only use a linear regression model for numerical estimation and a logistic model for categorical estimation.